submit_cgap/tests/test_submission.py

```diff
@@ -1202,3 +1212,471 @@ def mysterious_error(*args, **kwargs):
+
+    final_res = {
+        'submission_id': SOME_UUID,
+        "additional_data": {
+            "validation_output": [],
+            "post_output": [],
+            "upload_info": SOME_UPLOAD_INFO,
+        },
+        "processing_status": {
+            "state": "done",
+            "outcome": "success",
+            "progress": "irrelevant"
+        }
+    }
+
+    error_res = {
+        'submission_id': SOME_UUID,
+        'errors': [
+            "ouch"
+        ],
+        "additional_data": {
+            "validation_output": [],
+            "post_output": [],
+            "upload_info": SOME_UPLOAD_INFO,
+        },
+        "processing_status": {
+            "state": "done",
+            "outcome": "error",
+            "progress": "irrelevant"
+        }
+    }
+
+    def make_mocked_get(success=True, done_after_n_tries=1):
+        if success:
+            responses = (partial_res,) * (done_after_n_tries - 1) + (final_res,)
+        else:
+            responses = (partial_res,) * (done_after_n_tries - 1) + (error_res,)
+        response_maker = make_alternator(*responses)
+
+        def mocked_get(url, auth):
+            print("in mocked_get, url=", url, "auth=", auth)
+            assert auth == SOME_AUTH
+            if url.endswith("/me?format=json"):
+                return FakeResponse(200, json=make_user_record(
+                    project=SOME_PROJECT,
+                    user_institution=[
+                        {'@id': SOME_INSTITUTION}
+                    ]
+                ))
+            else:
+                assert url.endswith('/ingestion-submissions/' + SOME_UUID + "?format=json")
+                return FakeResponse(200, json=response_maker())
+        return mocked_get
+
+    mfs = MockFileSystem()
+
+    dt = ControlledTime()
+
+    with mock.patch("os.path.exists", mfs.exists):
+        with mock.patch("io.open", mfs.open):
+            with mock.patch.object(submission_module, "script_catch_errors", script_dont_catch_errors):
+                with mock.patch.object(submission_module, "resolve_server", return_value=SOME_SERVER):
+                    with mock.patch.object(submission_module, "yes_or_no", return_value=True):
+                        with mock.patch.object(submission_module, "get_keydict_for_server",
+                                               return_value=SOME_KEYDICT):
+                            with mock.patch.object(submission_module, "resolve_server", return_value=SOME_SERVER):
+                                with mock.patch.object(submission_module, "yes_or_no", return_value=True):
+                                    with mock.patch.object(submission_module, "get_keydict_for_server",
+                                                           return_value=SOME_KEYDICT):
+                                        with mock.patch("requests.post", mocked_post):
+                                            with mock.patch("requests.get", make_mocked_get(done_after_n_tries=3)):
```

Comment:
Although you can follow this if you reference each patch, some inline comments may be helpful in digesting this test.

Reply:
Especially since doing so might detect the duplication of `yes_or_no`. Yeah.
I also really want to explore doing this with decorators. A bit of this (the part that isn't dependent on a value from the definition environment) could be usefully done that way, reducing the indentation correspondingly

Reply:
I'll come back to this another day. I added a TODO comment.